File: backend/src/modules/buildings/building.controller.js
const buildingService = require('./building.service');
const siteVersionService = require('../siteVersions/siteVersion.service');

async function listBySite(req, res) {
  const buildings = await buildingService.listBuildingsBySite(
    req.params.siteId
  );
  res.json(buildings);
}

async function getById(req, res) {
  const building = await buildingService.getBuildingById(req.params.id);
  res.json(building);
}

async function createForSite(req, res) {
  const siteId = req.params.siteId;
  const building = await buildingService.createBuilding(siteId, req.body);
  await siteVersionService.syncWorkingPayloadFromDb(siteId);
  res.status(201).json(building);
}

async function update(req, res) {
  const building = await buildingService.updateBuilding(req.params.id, req.body);
  await siteVersionService.syncWorkingPayloadFromDb(building.siteId);
  res.json(building);
}

async function remove(req, res) {
  const deleted = await buildingService.deleteBuilding(req.params.id);
  await siteVersionService.syncWorkingPayloadFromDb(deleted.siteId);
  res.status(204).send();
}

module.exports = {
  listBySite,
  getById,
  createForSite,
  update,
  remove,
};

File: server/src/main/java/com/legioncontrols/server/sites/BuildingController.java
package com.legioncontrols.server.sites;

import com.legioncontrols.server.releases.ReleaseService;
import java.util.List;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

/**
 * Java equivalent of backend/src/modules/buildings/building.controller.js (routes split across
 * building.routes.js at /api/sites/:siteId/buildings and /api/buildings/:id in Express — combined
 * here since Spring's @RequestMapping doesn't need the same file-per-mount-point split).
 */
@RestController
public class BuildingController {

    private final BuildingService buildingService;
    private final ReleaseService releaseService;

    public BuildingController(BuildingService buildingService, ReleaseService releaseService) {
        this.buildingService = buildingService;
        this.releaseService = releaseService;
    }

    @GetMapping("/api/sites/{siteId}/buildings")
    public List<BuildingDto> listBySite(@PathVariable String siteId) {
        return buildingService.listBuildingsBySite(siteId);
    }

    @GetMapping("/api/buildings/{id}")
    public BuildingDto getById(@PathVariable String id) {
        return BuildingDto.from(buildingService.getBuildingRecordById(id));
    }

    @PostMapping("/api/sites/{siteId}/buildings")
    public ResponseEntity<BuildingDto> createForSite(@PathVariable String siteId, @RequestBody BuildingService.CreateBuildingRequest request) {
        var building = buildingService.createBuilding(siteId, request);
        releaseService.syncWorkingPayloadFromDb(siteId);
        return ResponseEntity.status(HttpStatus.CREATED).body(BuildingDto.from(building));
    }

    @PatchMapping("/api/buildings/{id}")
    public BuildingDto update(@PathVariable String id, @RequestBody BuildingService.UpdateBuildingRequest request) {
        var building = buildingService.updateBuilding(id, request);
        releaseService.syncWorkingPayloadFromDb(building.getSiteid());
        return BuildingDto.from(building);
    }

    @DeleteMapping("/api/buildings/{id}")
    public ResponseEntity<Void> remove(@PathVariable String id) {
        var deleted = buildingService.deleteBuilding(id);
        releaseService.syncWorkingPayloadFromDb(deleted.getSiteid());
        return ResponseEntity.noContent().build();
    }
}
